ratik/src/main/java/org/opencadc/ratik/InventoryValidator.java

```diff
@@ -150,7 +169,215 @@ public InventoryValidator(Map<String, Object> daoConfig, URI resourceID, Artifac
+                    }
+                    if (remoteArtifact == null) {
+                        remoteArtifact = remoteIterator.hasNext() ? remoteIterator.next() : null;
+                    }
+                    // TODO sanity check? if either iterator has no results in the first loop, exit?
+                    if (localArtifact == null && remoteArtifact == null) {
+                        artifactsToValidate = false;
+                        continue;
+                    }
+                    log.debug(String.format("comparing Artifacts:\n local - %s\nremote - %s",
+                                            localArtifact, remoteArtifact));
+
+                    // check if Artifacts are the same, or if the local Artifact
+                    // precedes or follows the remote Artifact.
+                    int order = orderArtifacts(localArtifact, remoteArtifact);
+                    if (order == 0) {
+                        validate(localArtifact, remoteArtifact);
+                        localArtifact = null;
+                        remoteArtifact = null;
+                    } else if (order < 0) {
+                        validate(localArtifact, null);
+                        localArtifact = null;
+                    } else {
+                        validate(null, remoteArtifact);
+                        remoteArtifact = null;
+                    }
+                }
+            } catch (IOException e) {
+                log.error("Error closing iterator: " + e.getMessage());
+            }
+        }
+    }
+
+    /**
+     * Useful for overriding in tests.
+     *
+     * @param localArtifact the local Artifact.
+     * @param remoteArtifact the remote Artifact.
+     */
+    void validate(Artifact localArtifact, Artifact remoteArtifact) {
+        log.debug(String.format("validating:\n local - %s\nremote - %s", localArtifact, remoteArtifact));
+        artifactValidator.validate(localArtifact, remoteArtifact);
+    }
+
+    /**
+     * Order two Artifacts on the String representation of Artifact.uri.
+     * Must match the ordering of a postgresql ORDER BY ASC on Artifact.uri.
+     * - if local equals remote returns 0.
+     * - if local lexicographically precedes remote returns a negative value.
+     * - if local lexicographically follows remote returns a positive value.
+     */
+    int orderArtifacts(Artifact localArtifact, Artifact remoteArtifact) {
+        log.debug(String.format("order artifact uri's:\n local - %s\nremote - %s",
+                                localArtifact == null ? "null" : localArtifact.getURI(),
+                                remoteArtifact == null ? "null" : remoteArtifact.getURI()));
+        int result;
+        if (localArtifact == null) {
+            result = 1;
+        } else if (remoteArtifact == null) {
+            result = -1;
+        } else {
+            result = localArtifact.getURI().toString().compareTo(remoteArtifact.getURI().toString());
```

Comment:
URI implements Comparable so you don't need to convert to string

This method that is easily testable with just code: please add unit test for the 5 scenarios: 
```
a,null 
null,b 
a,a
a,b 
b,a
```

Since this method is able to order non-null vs null, you need to state whether null is before or after non-null. The common thing is for null to come after non-null but I think you are doing the opposite here... it doesn't really matter as long as the conditionals up at line ~224+ do the same thing, but since we do have examples of "NULLS LAST" ordering elsewhere it's something to think about. It must be explicit.

re: comment about "ORDER BY uri ASC": this comment from cadc-inventory-db library when ordering by StorageLocation is related: 
```
// NOTE: StorageLocation.compare() specifies the correct order
// null storageBucket to come after non-null
// postgresql: the default order is equivalent to explicitly specifying ASC NULLS LAST
// default behaviour may not be db-agnostic
```

aside: if you want the string form of a URI (eg to serialise), always use toASCIIString(). toString() is a default method for debugging

Reply:
The null vs non-null behaviour was backwards to how it is handled else where. Updated to follow the same pattern where null follows non-null.
Added a unit test for the Artifact ordering method.